### ISSUE TITLE
FirstOrDefault after custom projection == null cannot be translated

### ISSUE THREAD

**Issue by groege** [dotnet/efcore#18476]
### Steps to reproduce
```
var queryA = db.MyEntity.Select(x => new MyEntityModel {
    ModelPropA = x.Children.Where(y => y.Type == Types.A).FirstOrDefault(),
    ModelPropB = x.Children.Where(y => y.Type == Types.B).FirstOrDefault(),
});
var queryB = queryA.Select(x => new MyEntityModel {
    ModelPropA = x.ModelPropA,
});

queryA.ToList() //This is working
queryB.ToList() //Throws System.InvalidOperationException: 'Sequence contains more than one element'

//EDIT:
queryA.Where(x => x.ModelPropA == null).ToList() //Throws the System.InvalidOperationException as well
```
[StackTrace](https://github.com/aspnet/EntityFrameworkCore/files/3750663/stacktrace.txt)

### Further technical details

EF Core version:
Database provider: Npgsql.EntityFrameworkCore.PostgreSQL 3.0.0
Target framework: NET Core 3.0
Operating system: WIN10
IDE: Visual Studio 2019 16.4.0 Preview 1.0

**Comment by ajcvickers** [comment]
@groege Please post a small, runnable project/solution or complete code listing that demonstrates the issue. This will ensure that we can reproduce what you are seeing.

**Comment by groege** [comment]
[ExceptionRepro.zip](https://github.com/aspnet/EntityFrameworkCore/files/3754313/ExceptionRepro.zip)

So I apologize - queryB actually works - I did have an error in my code (Which I found by doing the repro).

But queryA.Where(x => x.ModelPropA == null).ToList() is not working, it sais it cannot be translated.
See queryC in the repro project.

Thanks for your time!!

**Comment by smitpatel** [comment]
Simpler query
```C#
var query = db.Blogs.Where(e => db.Blogs.Select(b => new { b.Id }).FirstOrDefault() == null).ToList();
```

**Comment by groege** [comment]
@smitpatel 
That query doesn't seem simpler to me ^^ - looks convoluted to me (compared to just using a normal where clause)

**Comment by smitpatel** [comment]
@groege - Your normal where clause has custom projection preceding it. The simpler query is for repro and not a work-around.

**Comment by smitpatel** [comment]
This is anonymous type comparison.
Duplicate of #14672

**Comment by smitpatel** [comment]
Notes:
- If the result type of FirstOrDefault is non-nullable type then == null is always false (& inverse for != null)
- For nullable result type result of OrDefault family will be null only if the sequence is null. So we can rewrite this check to Any on sequence itself. For != null check, SingleOrDefault could twist things but following our philosophy or server eval, we treat it same as FirstOrDefault in subquery.
- This could also help us in translating entity equality with subquery & composite key (if we are not doing already)

**Comment by smitpatel** [comment]
> For nullable result type result of OrDefault family will be null only if the sequence is null. So we can rewrite this check to Any on sequence itself. For != null check, SingleOrDefault could twist things but following our philosophy or server eval, we treat it same as FirstOrDefault in subquery.

**Comment by smitpatel** [comment]
Re-activating as we didn't really fix this issue.

**Comment by smitpatel** [comment]
@ajcvickers - While this is not really patch issue, it is getting fixed in 6.0 branch as fixing bug for #26744 Not sure how to set milestone on this one now.
